plugins/formal-verification/src/org/jetbrains/kotlin/formver/conversion/StmtConversionVisitor.kt

```diff
@@ -70,36 +70,46 @@ class StmtConversionVisitor : FirVisitor<Exp, StmtConversionContext>() {
-            // When expressions with more than 2 branches can be embedded as nested if-else chains.
-            TODO("When expressions with more that two branches are not yet implemented")
-        }
+    override fun visitWhenSubjectExpression(whenSubjectExpression: FirWhenSubjectExpression, data: StmtConversionContext): Exp {
+        return whenSubjectExpression.whenRef.value.subject?.accept(this, data)
+            ?: throw Exception("FirWhenSubjectExpression $whenSubjectExpression has a null subject")
+    }
 
-        return when (whenExpression.subject) {
-            null -> {
-                val cvar = data.newAnonVar(data.convertType(whenExpression.typeRef.coneTypeOrNull!!))
-                val cond = whenExpression.branches[0].condition.accept(this, data)
-                // TODO: tidy this up and split it into helper functions.
-                val thenCtx = StmtConverter(data)
-                val thenResult = whenExpression.branches[0].result.accept(this, thenCtx)
-                thenCtx.addStatement(Stmt.LocalVarAssign(cvar.toLocalVar(), thenResult))
-                val elseCtx = StmtConverter(data)
-                val elseResult = whenExpression.branches[1].result.accept(this, elseCtx)
-                elseCtx.addStatement(Stmt.LocalVarAssign(cvar.toLocalVar(), elseResult))
-                data.addDeclaration(cvar.toLocalVarDecl())
-                data.addStatement(
-                    Stmt.If(
-                        cond,
-                        thenCtx.block,
-                        elseCtx.block
-                    )
-                )
-                cvar.toLocalVar()
+    private fun visitWhenBranches(whenBranches: List<FirWhenBranch>, data: StmtConversionContext, cvar: ConvertedVar): Exp {
+        // NOTE: I think that this will also work with "in" or "is" conditions when implemented, but I'm not 100% sure
+        val cond = whenBranches[0].condition.accept(this, data)
+        val thenCtx = StmtConverter(data)
+        val thenResult = whenBranches[0].result.accept(this, thenCtx)
+        thenCtx.addStatement(Stmt.LocalVarAssign(cvar.toLocalVar(), thenResult))
+        val elseBranch =
+            when {
+                // When last branch is not an else
+                whenBranches.size == 1 -> Stmt.Seqn(listOf(), listOf())
+                // When last branch is an else
+                whenBranches.size == 2 && whenBranches[1].condition is FirElseIfTrueCondition -> {
+                    val elseCtx = StmtConverter(data)
+                    val elseResult = whenBranches[1].result.accept(this, elseCtx)
+                    elseCtx.addStatement(Stmt.LocalVarAssign(cvar.toLocalVar(), elseResult))
+                    elseCtx.block
+                }
+                // Recursive case
+                else -> {
+                    val elseCtx = StmtConverter(data)
+                    visitWhenBranches(whenBranches.drop(1), elseCtx, cvar)
```

Comment:
The recursion here is fine, but I feel like it should be possible to handle this with the empty list as the base case instead of the singleton list.  WDYT?